packages/lib/gpc/src/gpcCompile.ts

```diff
@@ -751,27 +770,33 @@ function compileProofEntryConstraints(
     entryEqualToOtherEntryByIndex: CircuitSignal[];
   };
   entryConstraintMetadata: {
-    firstOwnerIndex?: number;
+    firstOwnerIndex: Record<IdentityProtocol, number | undefined>;
   };
 } {
   // Deal with equality comparision and POD ownership, which share circuitry.
-  let firstOwnerIndex = undefined;
+  const firstOwnerIndex = Object.fromEntries(
+    [SEMAPHORE_V3, SEMAPHORE_V4].map((idType) => [idType, undefined])
+  ) as Record<IdentityProtocol, number | undefined>;
   const entryEqualToOtherEntryByIndex: bigint[] = [];
   const virtualEntryEqualToOtherEntryByIndex: bigint[] = [];
 
   for (const entryInfo of entryMap.values()) {
-    // An entry is always compared either to the first owner entry (to ensure
-    // only one owner), or to another entry specified by config, or to itself
-    // in order to make the constraint a nop.
-    if (entryInfo.entryConfig.isOwnerID) {
-      if (firstOwnerIndex === undefined) {
-        firstOwnerIndex = entryInfo.entryIndex;
+    // An entry is always compared either to the first owner entry of each
+    // supported identity type (to ensure only one owner with respect to that
+    // type), or to another entry specified by config, or to itself in order to
+    // make the constraint a nop.
+    const ownerIDType = entryInfo.entryConfig.isOwnerID;
+    if (ownerIDType !== undefined) {
+      if (firstOwnerIndex[ownerIDType] === undefined) {
+        firstOwnerIndex[ownerIDType] = entryInfo.entryIndex;
       } else if (entryInfo.entryConfig.equalsEntry !== undefined) {
         throw new Error(
           "Can't use isOwnerID and equalsEntry on the same entry."
         );
       }
-      entryEqualToOtherEntryByIndex.push(BigInt(firstOwnerIndex));
+      entryEqualToOtherEntryByIndex.push(
+        BigInt(firstOwnerIndex[ownerIDType] as number)
```

Comment:
Interesting.  Can the compiler not figure out that this value can't be undefined because of the check above?  Does it assume it could've changed between the two lookups?  You might be able to avoid the extra cast if you save it to a local variable first `const relevantOwnerIndex = firstOwnerIndex[ownerIDType]`

Reply:
I think it assumes that since `firstOwnerIndex : Record<IdentityProtocol, number | undefined>`, `firstOwnerIndex[i] : number | undefined` must always hold.